tools/clang/lib/SPIRV/EmitVisitor.cpp

```diff
@@ -1026,5 +1091,401 @@ uint32_t EmitTypeHandler::emitType(const SpirvType *type,
+    } else {
+      offset = roundToPow2(offset, memberAlignment);
+    }
+
+    // The vk::offset attribute takes precedence over all.
+    if (vkOffsetAttr) {
+      offset = vkOffsetAttr->getOffset();
+    }
+    // The :packoffset() annotation takes precedence over normal layout
+    // calculation.
+    else if (packOffsetAttr) {
+      const uint32_t packOffset = packOffsetAttr->Subcomponent * 16 +
+                                  packOffsetAttr->ComponentOffset * 4;
+      // Do minimal check to make sure the offset specified by packoffset does
+      // not cause overlap.
+      if (packOffset < nextLoc) {
+        emitError("packoffset caused overlap with previous members",
+                  packOffsetAttr->Loc);
+      } else {
+        offset = packOffset;
+      }
+    }
+
+    // Each structure-type member must have an Offset Decoration.
+    emitDecoration(typeResultId, spv::Decoration::Offset, {offset}, index);
+    offset += memberSize;
+
+    // Each structure-type member that is a matrix or array-of-matrices must be
+    // decorated with
+    // * A MatrixStride decoration, and
+    // * one of the RowMajor or ColMajor Decorations.
+    if (auto *arrayType = dyn_cast<ArrayType>(fieldType)) {
+      // We have an array of matrices as a field, we need to decorate
+      // MatrixStride on the field. So skip possible arrays here.
+      fieldType = arrayType->getElementType();
+    }
+
+    // Non-floating point matrices are represented as arrays of vectors, and
+    // therefore ColMajor and RowMajor decorations should not be applied to
+    // them.
+    if (auto *matType = dyn_cast<MatrixType>(fieldType)) {
+      if (isa<FloatType>(matType->getElementType())) {
+        memberAlignment = memberSize = stride = 0;
+        std::tie(memberAlignment, memberSize) =
+            getAlignmentAndSize(fieldType, rule, &stride);
+
+        emitDecoration(typeResultId, spv::Decoration::MatrixStride, {stride},
+                       index);
+
+        // We need to swap the RowMajor and ColMajor decorations since HLSL
+        // matrices are conceptually row-major while SPIR-V are conceptually
+        // column-major.
+        if (matType->isRowMajorMat()) {
+          emitDecoration(typeResultId, spv::Decoration::ColMajor, {}, index);
+        } else {
+          // If the source code has neither row_major nor column_major
+          // annotated, it should be treated as column_major since that's the
+          // default.
+          emitDecoration(typeResultId, spv::Decoration::RowMajor, {}, index);
+        }
+      }
+    }
+
+    ++index;
+  }
+}
+
+void EmitTypeHandler::emitDecoration(uint32_t typeResultId,
+                                     spv::Decoration decoration,
+                                     llvm::ArrayRef<uint32_t> decorationParams,
+                                     uint32_t memberIndex) {
+
+  spv::Op op = memberIndex ? spv::Op::OpMemberDecorate : spv::Op::OpDecorate;
```

Comment:
Better to assert `curDecorationInst` is empty before `.clear()` in case we messed something up.

Reply:
If it is empty, there would be no point in clearing it. It is not going to be empty as it will contain the last decoration we had created. I've changed the code to clear it at the end of this function, and also assert in the beginning.